scripts/semeval2017_to_coll2003_and_relations.py

```diff
@@ -0,0 +1,177 @@
+"""
+
+Input:
+    Text:
+        Based on the theoretical analysis, the value of the measuring resistor...
+
+    Annotations:
+        T1	Process 39 70	value of the measuring resistor
+        T2	Process 72 74	Rm
+        *	Synonym-of T2 T1
+        T3	Material 427 466	NI-USB 6009 analog-to-digital converter
+        T4	Material 322 336	4.7kΩ resistor
+        T5	Material 375 402	saturated calomel electrode
+        T6	Process 472 500	electrochemical noise signal
+        T7	Material 520 547	in-house developed software
+        T8	Material 778 785	dataset
+        T10	Task 674 708	obtain a single value of potential
+        T11	Task 979 1051	accurate recording of the potential noise in the frequencies of interest
+        T12	Process 97 114	corrosion process
+        T13	Process 136 161	value of noise resistance
+        T14	Task 175 199	validate this conclusion
+        T15	Material 257 294	pair of nominally identical specimens
+        T16	Task 13 33	theoretical analysis
+
+Output 1:  Entity extraction (CONLL2003 format)
+
+
+Output 1:  Relation classification
+    JSONlines where each JSON is:
+    {
+        "text": "Based on the theoretical analysis , the [[ value of the measuring resistor ]] ...",
+        "label": "Synonym-of",
+        "metadata": [7, 11, 12, 12]
+    }
+    ** only contains sentences with relations; entities are gold wrt their Arg1/Arg2 position; no cross-sentence relations **
+
+Output 2:  Relation extraction
+    ** same format as above, but includes sentences without relations & all entity pairs are attempted as Arg1/Arg2 positions **
+
+
+
+Pseudocode for 1:
+    - sentence-ify & tokenize text & assign char-level span indices to each token
+    - for each sentence:
+        for each token:
+            if within entity, do stuff
+
+
+"""
+
+from typing import *
+
+import os
+import spacy
+nlp = spacy.load('en_core_web_md')
+
+from sci_bert.common.span import Span, TokenSpan, MentionSpan, label_sent_token_spans
+
+MIN_NUM_TOKENS_PER_SENT = 5
+
+os.makedirs('data/ner/semeval2017/', exist_ok=True)
+for split in ['train', 'dev', 'test']:
+    with open(f'data/ner/semeval2017/{split}.txt', 'w') as f_out:
+        ann_dir = f'semeval2017/{split}/'
+        instance_ids = sorted({os.path.splitext(ann_file)[0] for ann_file in os.listdir(ann_dir)})
+        for id in instance_ids:
+            ann_file = os.path.join(ann_dir, f'{id}.ann')
+            txt_file = ann_file.replace('.ann', '.txt')
+
+            # load text
+            with open(txt_file, 'r') as f_txt:
+                text = f_txt.read().strip()
+                text = ''.join([char if char.strip() != '' else ' ' for char in text])  # normalize whitespaces, such as '\xa0' --> ' '
+                spacy_text = nlp(text)
+
+            # split sentences & tokenize
+            sent_token_spans = TokenSpan.find_sent_token_spans(text=text, sent_tokens=[[token.text for sent in spacy_text.sents for token in sent if token.text.strip() != '']])
```

Comment:
BERT will probably complain about long sequences (entire abstract). So we need to do this at sentence level.